### ISSUE TITLE
Include Project/Assembly in the Symbol Tooltip and Member Filter

### ISSUE THREAD

**Issue by wmjordan** [wmjordan/Codist#150]
As @fitdev's suggestion in #67 

> A few more suggestions:
> - Perhaps you can include Assembly/Project name/path (whichever is easier for you) for types as well (in addition to including that info for namespaces)?
> - The Files popup that appears when you click on a namepsace in the dropdown is great! Maybe you can include the Assembly/Project name for the source file tooltips as well?
> - And in conjunction with 2, a way to filter source files based on Project/Assembly would be even better (similar to how you have filters for the kinds of types for example). I realize it may require some work as you would need to include a combobox with a list of applicable projects/assemblies.

**Comment by wmjordan** [comment]
@fitdev
> * Maybe you can include the Assembly/Project name for the source file

I found that this can be an O(N) operation. If a solution contains many files, that can take some time to check each file in all projects one by one against the source file location.

**Comment by wmjordan** [comment]
Please try the new beta which has implemented the 1st and 2nd requests.

**Comment by fitdev** [comment]
Works great! Thank you for fast and excellent work! Also I see that you have fixed an issue where now the dropdown displays types and namespaces from _all projects_ (not just from the one to which currently active file belongs).

**Comment by wmjordan** [comment]
Thank you for your feedback.

> an issue where now the dropdown displays types and namespaces from all projects (not just from the one to which currently active file belongs)

I don't know whether this is fully fixed yet.

Given that there are three projects, _P_, _P.A_, _P.B_. Project _P.A_ depends on _P_, _P.B_ depends on _P_ too. But neither _P.A_ nor _P.B_ references each other. If we open a code file in _P.A_, the NaviBar displays the namespaces _P_, _A_. In the menu for _P_, only namespaces and types in _P_ and _P.A_ are listed. Things in _P.B_ are not displayed. The same thing can happen to code files in project _P.B_.

**Comment by fitdev** [comment]
You are right! That's a good catch! I can confirm this behavior.

Same goes for opening code file from _P_. Then neither types from _P.A_, nor from _P.B_ are shown.

I guess this is rather complicated since Roslyn is per-project-scoped apparently. Maintaining a list of loaded projects and for-eaching every one of them for a given namespace maybe too expensive and/or complex. So maybe these scenarios you can add to a backlog (and have this extended "search" be togglable with an option in case it has big performance impact)? I am already quite happy with the way it works currently.

**Comment by fitdev** [comment]
To expand a bit on the 4 possible scenarios...

We have:
1. Namespaces and types that belong to the current Project (for which we are in the opened code file).
2. Namespaces and types that belong to _other projects_, which the current project _depends upon_.

More complex scenarios not addressed:
3. Namespaces and types that belong to _dependent_ projects (those that depend on the current project).
4. Namespaces and types that have no dependency relation to the current project, but have the same namespace hierarchy.

So, in that regard my suggestions would be:
- For 3 and 4, if it is in any way difficult, postpone it and put it into backlog.
- 3 and 4, if involve a performance hit could be available only if a certain option is enabled.
- For case 2 (and 3 and 4 if/when they are implemented): color entries slightly differently to indicate they come from other projects (maybe a bit more transparent (so that it retains all the original colors but still looks different))
- At some point in the future, if 3 and 4 are realized have a filter / dropdown / toolbar that lets one select items from which of the 4 categories will be shown.

**Comment by wmjordan** [comment]
It is possible to address those namespaces in 3 and 4, by getting the `Solution`, then for each `Project`s that are not yet included in 1 or 2 (OMG, I have to figure out what projects are included in 1 and 2), get a `Compilation`, then get `GlocalNamespace`, for each namespace members, match their names with the current namespace. If current namespace is hierarchical one, search for it hierarchically. Once we get those "sibling" namespaces, get their members.

It is not so difficult, but it really requires some works.

**Comment by fitdev** [comment]
I realize that, which is why I suggested maybe you put that into a backlog, as your valuable time would best be used to address other features. Also, I fear that the performance impact may be a bit too high.

**Comment by fitdev** [comment]
Latest beta 6164 seem to have broken the display of items from 2 (Namespaces and types that belong to other projects, which the current project depends upon). It just shows namespaces and types from the current project only.

**Comment by wmjordan** [comment]
> Latest beta 6164 seem to have broken the display of items from 2

It should not.
Please take care about the Project drop-down on the left of the NaviBar and see what project setting is active.

**Comment by fitdev** [comment]
What do you mean? There is just one item (in most cases) to the left of the NaviBar. Namely it's the project that the currently opened document belongs to (unless it happens to be a shared file, which I have not tried).

6155 used to display items from _both_:
- the current project (i.e. the one to which the currently opened file belongs)
- and _all_ projects on which the current one _depends on_

6164 now displays just the items from the current project.

**Comment by fitdev** [comment]
UPDATE: Now that I have tested a few more files, it works properly on the _newly opened files_, but on the tabs/files where it did no work, when switching to them it _still does not work_. Only closing and reopening a document fixes it. I guess it has to do with me trying it out before the solution was fully loaded, and then the results were simply cached.

**Comment by wmjordan** [comment]
> it has to do with me trying it out before the solution was fully loaded, and then the results were simply cached

Hmm, it seems to be the cause.



> I do think that 3 and 4 are valuable and useful. It's just that I am not sure of the performance impact and given that they require a lot of work that these features should be done soon.

I think that is useful too. I am trying to implement them.
Performance may not be the major problem at this moment.
The obstacle is to avoid redundant items showing up in the drop-down.
Since both _P.A_ and _P.B_ have referenced _P_, while enumerating the projects. Items in _P_ will show up at least 3 times when they are queried. Once in _P_ itself, twice in _P.A_ and _P.B_.

**Comment by fitdev** [comment]
> Please try clicking somewhere else outside the namespace scope and then inside to refresh the NaviBar.

This almost fixed it. But in addition I had to open another document and then switch back. Then it displayed the missing items.

> Since both P.A and P.B have referenced P, while enumerating the projects. Items in P will show up at least 3 times when they are queried. Once in P itself, twice in P.A and P.B.

For the case where we want to show items from _all_ projects, wouldn't it then be simpler to scan _all_ projects one by one? That way every item will appear only once.

**Comment by wmjordan** [comment]
Please try the new beta which has implemented aforementioned 3, 4.

**Comment by wmjordan** [comment]
> For the case where we want to show items from _all_ projects, wouldn't it then be simpler to scan _all_ projects one by one? That way every item will appear only once.

That was what I thought. However, things in Roslyn are not so simple.

**Comment by fitdev** [comment]
For 3 I get lots of duplicates. So what you have described previously is happening. Basically all the items are duplicated - one copy per dependent project I guess.

> That was what I thought. However, things in Roslyn are not so simple.

Well, you know best. But I just assumed that if, for the moment, we do not care about showing / tracking project inter-dependencies, but simply want to show all user-code-defined namespaces and types, we could simply scan all opened projects in a solution. And then for each project query it via Roslyn to actually get namespaces and types.

**Comment by wmjordan** [comment]
> For 3 I get lots of duplicates.

I got them too not long after publishing the beta.



> for each project query it via Roslyn to actually get namespaces and types.

The problem was that in _P_ you have a namespace `P` and in _P.A_ you have another namespace `P`--they have the same name, but they don't equal to each other.

Anyway, please help test the new beta.

**Comment by fitdev** [comment]
After I tested 6184:

- Now there are no duplicates, but instead, depending on where the cursor is in the document I get 1 (correct), 2, or 3 identical root namespaces on the NaviBar itself, i.e. it shows something like: `P > P > P > MyOpenedType`.
- Types that are listed in the _popups_ that appear when you click on the namespace are missing (i.e. only the types that are defined in the _active project_ are shown, and types defined in the same namespace but a different library are not shown).

Also, there do appear to be duplicate namespace entries as well, but this, I think may be related to what we mean by namespaces being the same, which is why I was very surprised when you told me that namespaces can have the same name but not equal to each other:

> The problem was that in P you have a namespace P and in P.A you have another namespace P--those two namespaces have the same name, but they don't equal to each other.

To me it would make sense to consider those "2" namespaces the same namespace. I mean, if both namespaces are `global::P`, but one is defined in a project `P.dll` and another is defined in `AnotherP.dll`, _even if these 2 projects do not depend on one another_, I think it still makes sense to consider the _namespaces_ to be the same! So, I don;t think there should ever be a situation when there are duplicate _namespace_ entries, since we would use strictly string comparison only using C# / .Net rules (i.e. case-sensitive comparison).

The situation is different however with respect to types. Assuming that we have `P.TypeA` in `P.dll` and `P.TypeA` in 'AnotherP.dll`, and both projects are loaded, then we _should see 2 entries_ for the _type_ `TypeA`, both under `P` namespace (namespace is considered one and the same despite being used in 2 unrelated projects).

So, given that, at least to me, 2 namespaces should be considered the same if it's their string names are equal, can't the code be simplified to basically for-eaching over all projects and all namespaces / types and saving results in something like: `Dictionary<string /* Namespace full name */, (TypeRef, ProjectRef)>`?

**Comment by wmjordan** [comment]
> * Now there are no duplicates, but instead, depending on where the cursor is in the document I get 1 (correct), 2, or 3 identical root namespaces on the NaviBar itself, i.e. it shows something like: `P > P > P > MyOpenedType`.
> * Types that are listed in the _popups_ that appear when you click on the namespace are missing (i.e. only the types that are defined in the _active project_ are shown, and types defined in the same namespace but a different library are not shown).

Thank you for your tests.
I found and managed to fix the first issue in the new beta.

However, I could not reproduce the second issue.
Could you make a small solution which reproduce this for me to test?

**Comment by wmjordan** [comment]
> I was very surprised when you told me that namespaces can have the same name but not equal to each other

This not only happens to namespaces, but also types and members defined within those namespaces!
If there's a .NET Core project, which allows you to target multiple platforms, for instance, .NET 4, .NET 5, .NET Core 3.1. And there is a class named _A_ in a code file. You eventually get 3 classes _A_ if you enumerate all projects within the solution!

If there is another .NET Core project, which also targets the above 3 platforms, the project does not define a new class named _A_, but references the previous project.
Ultimately you get 3+3=6 classes _A_ if you enumerate all projects within the solution! Each of those 6 classes does not equal to each other and there is no O(1) operation to determine their equality.

**Comment by fitdev** [comment]
I can confirm that `P > P > P > MyOpenedType` is fixed in 6188.

Regarding another issue, the situation is the following:

- _Some_ namespaces are duplicated (one namespace copy per Project) I guess this is by design currently, but perhaps we can discuss how best to present it (merge into one item, or have several items, but in parentheses or elsewhere provide Project's name).

- However, within the same dropdown other namespaces are _not_ duplicated (even though they should be - according to the current design - since they are defined in _multiple_ projects). Strangely, the _types from those other projects_ that belong to the same namespace for which dropdown is shown _are present in the dropdown_ (even though child namespaces from those other projects are not).

So, assuming that there are:
`A.Type1` in `A.dll`
`A.B` namespace with some types in `A.dll`
`A.Type2` in `A.B.dll` (A.B.dll references A.dll)
`A.B` namespace with some types in `A.B.dll` (A.B.dll references A.dll)
`A.C` namespace with some types in `A.C.dll` (A.C.dll references A.dll)
`A.C` namespace with some types in `A.C.D.dll` (A.C.D.dll references A.dll and A.C.dll)
and I am currently viewing `A.Type1` (active project is `A.dll`), and I dropdown the `A` namespace node:

1. `A.Type2` from `A.B.dll` is _shown_ under `A` directly, which is correct
2. `A.B` namespace from `A.dll` _is_ shown, which is correct
3. `A.B` namespace from `A.B.dll` is _not_ shown, which is _wrong_ (should be shown)
4. `A.C` namespace from `A.C.dll` _is_ shown, which is correct
5. `A.C` namespace from `A.C.D.dll` _is_ shown, which is correct

I have no idea what the difference between 3 and 4 is. They should behave the same, but for some reason, in one case namespaces are duplicated, and in another case namespace from _depending_ project is not shown (even though types from the same project are shown).

**Comment by fitdev** [comment]
And regarding then the presentation of duplicate entries, there could be 2 possibilities:

- Merge namespaces with identical names into one item
- Have one namespace item per each project where such a namespace is defined

I think I prefer the first option for namespaces.

For types, however, I think there should definitely be _distinct_ entries for each type with the same name (and namespace) but from different projects. I mean this is a rare case, but it is possible I suppose.

Multitargeting, on the other hand, does seem to make things much more complicated. I did not realize that with multitargeting VS basically doubles (multiplies) the number of projects. I don't use it, and right now cannot think of the best way to deal with it. I suppose, if you can do it without too much work, items that are the same (i.e. full names match and they are from the same "real" project) should be merged into just 1 item.

**Comment by wmjordan** [comment]
> 3\. `A.B` namespace from `A.B.dll` is _not_ shown, which is _wrong_ (should be shown)

I still could not reproduce this.
I have a project _A_ referenced by _A.X_, _A.Y_ and _A.Z_. When I click the `A` namespace node from a file defined in _A_, the drop down contains sub-namespaces and types under `A` from all four projects.

**Comment by fitdev** [comment]
That's strange. But I guess it's not consistent on my end. Meaning that 3 does not apply to _all_ subnamespaces `A.B.dll` but apparently only to those subnamespaces that are _also defined in_ `A.dll`. So that if `A.dll` has `A.B` namespace, and `A.B.dll` (that depends on `A.dll`) has `A.B` namespace as well (with different types than those defined in `A.dll`). Then the `A.B` namespace _from_ `A.B.dll` is _not_ displayed from an active file defined in `A.dll`. However `A.C` namespace defined _only_ in `A.B.dll` (i.e. _not_ present in `A.dll`) _is shown_ in the `A` dropdown from an active file defined in `A.dll`.


On-to a different topic. Just a quick suggestion... Right now when you click on a namespace in the dropdown, it opens a popup with a list of subnamespaces and types. That's great! But then, if you click once more on a _snamespace_ item _in that popup_, the original popup is closed and a new popup is shown. But sometimes it is useful to leave the original popup open still even when you are openning additional popup by clicking on the child namespaces. So, perhaps if you do `Ctrl+Click` instead of regular click it will _not_ hide the previous popup, i.e. so that it behaves more like a menu.

**Comment by wmjordan** [comment]
> if `A.dll` has `A.B` namespace, and `A.B.dll` (that depends on `A.dll`) has `A.B` namespace as well (with different types than those defined in `A.dll`). Then the `A.B` namespace _from_ `A.B.dll` is _not_ displayed from an active file defined in `A.dll`

No, there will not be two _A_B_ namespace in the drop down menu.
They will be merged into one.

**Comment by fitdev** [comment]
> No, there will not be two A_B namespaces in the drop down menu.
> They will be merged into one.

That's even better! But in any case, it does not happen either. The types from `A.B` namespace defined in `A.B.dll` are _not shown_ in the displayed `A.B` namespace (`B` namespace item in the dropdown) where only types defined in `A.dll` are shown.

**Comment by wmjordan** [comment]
OMG, I could not reproduce this.
Could you take the trouble of making a mini solution which can reproduce your problem?

**Comment by fitdev** [comment]
Here's the repro.
Open `ClassInA1.cs` in Project `A`. Then on the NaviBar, click on `A` namespace element. It will show 3 items in the dropdown: `B` namespace, `C` namespace, and `ClassInA1` type. Now click on the `B` namespace subitem. The new popup lists _just_ `ClassInA_B` type, which is defined in `A.dll`. However, the type `A.B.ClassInB1` defined in `B.dll` is _not_ listed.

**Comment by wmjordan** [comment]
Thank you for the demo.
The inconsistency was because of different algorithm was used to list namespace members.

Please try the new beta.

**Comment by fitdev** [comment]
Thank you for the new beta! Glad you were able to reproduce the issue. Yes, now this seems to be fixed! Thank you!

There are 3 small issues:

1. Why there are still some duplicated / unmerged namespaces in the dropdown? Shouldn't all identically named namespaces be merged?
2. I see you have implemented `Ctrl+Click` behavior to show/hide the "parent" menu. However this seems to apply only to the drodown _menu_, and _not_ to the popup window that appears when you click on the namespace in the dropdown. The same behavior should _also_ apply to the _popup windows_. So that: 

- NaviBar.Namespace.Click: Shows dropdown with subnamespaces and types
- Dropdown.Namespace.CtrlClick: Shows a _popup_ with a list of subnamespaces and types, the _dropdown_ stays open (because of Ctrl+Click)
- Popup.Namespace.CtrlClick: Shows a _second popup_ with a list of subsubnamespaces and types, both the original dropdown and _the "parent" popup_ stay open (because of Ctrl+Click)

So the last item is not working currently. The "parent" popup closes regardless of Ctrl+ or regular Click.
3. In case of a large solution / many namespaces / types, it takes a while (10+ seconds) for a NavIBar's dropdown to be shown. Most often it would just be empty, or sometimes nothing would happen on click. Perhaps some kind of UI cue could be added to indicate to the user that NaviBar is gathering information across the solution and preparing a list of items to display. Currently it will _eventually_ be displayed after several click attempts.

**Comment by wmjordan** [comment]
> The same behavior should also apply to the popup windows.

Yes, I will check to make sure that and they will be closed if items inside are activated via keyboard by @fantasyz's request--I did not realized I'd modified the wrong piece of code until today, thus his request were acknowledged but not implemented at all.

> In case of a large solution / many namespaces / types, it takes a while (10+ seconds) for a NavIBar's dropdown to be shown. Most often it would just be empty, or sometimes nothing would happen on click.

This is really a problem...
How large is your largest solution currently?
Mine has about 1000 code files, but it does not take so long to display the drop-down.

**Comment by fitdev** [comment]
I mostly work on just one solution. It's around 60 projects, 7000 files, and maybe 8000 types.

**Comment by fitdev** [comment]
Also, another small thing:

4. In the tooltip for namespace items, it says "Location: [number]". Is that supposed to be number of code files where the namespace declaration is defined? Does it include `using`s? The numbers seem way too high for me (unless it also includes all the `using`s).

**Comment by fitdev** [comment]
5. And one more thing - got a VS crash after: right-click on a namespace item in the NaviBar's dropdown, click on Select Code (new document opens), close the newly opened docuemnt and return back to the original document with the dropdown still open, VS Crash:

```
Application: devenv.exe
Framework Version: v4.0.30319
Description: The process was terminated due to an unhandled exception.
Exception Info: System.NullReferenceException
   at Codist.ToolTipFactory+<>c__DisplayClass0_0.<CreateToolTip>b__1(Microsoft.CodeAnalysis.SyntaxReference)
   at System.Linq.Enumerable+WhereSelectArrayIterator`2[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089],[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].MoveNext()
   at System.Linq.Enumerable+<DistinctIterator>d__64`1[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].MoveNext()
   at System.Linq.Enumerable+WhereSelectEnumerableIterator`2[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089],[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].MoveNext()
   at System.String.Join(System.String, System.Collections.Generic.IEnumerable`1<System.String>)
   at Codist.ToolTipFactory.CreateToolTip(Microsoft.CodeAnalysis.ISymbol, Boolean, Codist.SemanticContext)
   at Codist.Controls.SymbolList+<CreateItemToolTipAsync>d__47.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(System.Threading.Tasks.Task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(System.Threading.Tasks.Task)
   at Codist.Controls.SymbolList+<ShowToolTipForItemAsync>d__46.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(System.Threading.Tasks.Task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(System.Threading.Tasks.Task)
   at Codist.Controls.SymbolList+<MouseMove_ChangeToolTip>d__44.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore+<>c.<ThrowAsync>b__6_0(System.Object)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate, System.Object, Int32)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(System.Object, System.Delegate, System.Object, Int32, System.Delegate)
   at System.Windows.Threading.DispatcherOperation.InvokeImpl()
   at System.Windows.Threading.DispatcherOperation.InvokeInSecurityContext(System.Object)
   at MS.Internal.CulturePreservingExecutionContext.CallbackWrapper(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at MS.Internal.CulturePreservingExecutionContext.Run(MS.Internal.CulturePreservingExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Windows.Threading.DispatcherOperation.Invoke()
   at System.Windows.Threading.Dispatcher.ProcessQueue()
   at System.Windows.Threading.Dispatcher.WndProcHook(IntPtr, Int32, IntPtr, IntPtr, Boolean ByRef)
   at MS.Win32.HwndWrapper.WndProc(IntPtr, Int32, IntPtr, IntPtr, Boolean ByRef)
   at MS.Win32.HwndSubclass.DispatcherCallbackOperation(System.Object)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate, System.Object, Int32)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(System.Object, System.Delegate, System.Object, Int32, System.Delegate)
   at System.Windows.Threading.Dispatcher.LegacyInvokeImpl(System.Windows.Threading.DispatcherPriority, System.TimeSpan, System.Delegate, System.Object, Int32)
   at MS.Win32.HwndSubclass.SubclassWndProc(IntPtr, Int32, IntPtr, IntPtr)
```

**Comment by wmjordan** [comment]
> 1. right-click on a namespace item in the NaviBar's dropdown, click on Select Code (new document opens), close the newly opened docuemnt and return back to the original document with the dropdown still open, VS Crash

What's the version number?
It should have been fixed in the most recent beta.

**Comment by wmjordan** [comment]
> I mostly work on just one solution. It's around 60 projects, 7000 files, and maybe 8000 types.
> 
> Though I should clarify that how long it takes for the dropdown to be shown depends on which namespace I click. Sometimes it is shown instantly - if the namespace is "small". If it is one of the root namespaces though, in a file in a project on which many other projects depend, then it takes a while.

Oh, it is pretty large one.
Yes, the show-up speed depends on the accumulated items of the comprised namespace. More files more members, slower.
Anyway, I am cloning the source code of Roslyn (a much bigger repository) and try to conquer the hell mode.

> 4. "Location: [number]". Is that supposed to be number of code files where the namespace declaration is defined? Does it include `using`s?

It does not include the `using`s. It should be number of the files which have the namespace declaration.
Notice that namespace declarations like `namespace A.B`, `namespace A.C` will make namespace `A` be counted twice, as well as `A.B` and `A.C` each once respectively.

> 5. And one more thing - got a VS crash

It should be fixed in the last beta.

**Comment by wmjordan** [comment]
> > 1. And one more thing - got a VS crash
> 
> It should be fixed in the last beta.

Well, in the hell mode (handling Roslyn's source code) Codist crashed again.
Debugging...

**Comment by wmjordan** [comment]
Hey friends, have your upgraded your VS to 16.9?
I met with some crashes. Please download the new beta which has addressed them.

**Comment by fantasyz** [comment]
@wmjordan Not yet. I didn't get too much crashing from last build. Maybe at most ~1 in every two days.

**Comment by wmjordan** [comment]
> Maybe at most ~1 in every two days.

OMG, that's terrible.
Please post the stack trace if bad things happened and Codist was involved.

**Comment by fitdev** [comment]
> Hey friends, have your upgraded your VS to 16.9?
I have been using it for the last 2 months (previews) and experienced just one crash that I reported. Now I am onto 16.10 Preview 1 and have not seen crashes yet, but I did not test it thoroughly yet either.
